### ISSUE TITLE
Follow Best Pratices for Java Libraries

### ISSUE THREAD

**Issue by mjacoby** [eclipse-digitaltwin/aas4j#35]
Please, 

- [ ] review [Google Best Practices for Java Libraries](https://jlbp.dev) and
- [ ] evaluate where it should be applied and considered into our code

**Comment by mjacoby** [comment]
My take on how the Best Practices affect aas4j

| JLBP #  | Name | Effort | Notes |
| ------------- | ------------- | ------------- | ------------- |
| 1 | Minimize dependencies | low/medium |
| 2 | Minimize API surface | medium/high | For each class/method decide if public/final |
| 3 | Use Semantic Versioning | none/low | optimal: find tool to ensure compliance to semantic versioning |
| 4 | Avoid dependencies on unstable libraries and features |  none |
| 5 | Do not include a class in more than one classpath entry | none |
| 6 | Rename artifacts and packages together | none | when changing maven coordinates change namespace too (which is what we are going to do anyways) |
| 7 | Make breaking transitions easy | none | optional: guidelines how to do breaking changes (e.g. by using @Deprecated)
| 8 | Advance widely used functionality to a stable version | none |
| 9 | Support the minimum Java version of your consumers | low/none | Currently using Java 11 - but can we go lower?
| 10 | Maintain API stability as long as needed for consumers | none/low | Concept how to maintain old versions of meta model in the future
| 11 | Keep dependencies up to date | low/medium | Use Dependabot, plan for regular releases (as dependencies update rather often)
| 12 | Make level of support and API stability clear | low | 
| 13 | Remove references to deprecated features in dependencies at the first opportunity | ? | should be able to automatically scan for this, probably compiler gives a warning
| 14 | Specify a single, overridable version of each dependency | low |
| 15 | Publish a BOM for multi-module projects | ? | I don't fully understand this yet
| 16 | Ensure upper version alignment of dependencies for consumers | low |
| 17 | Coordinate rollout of breaking changes | none  |
| 18 | Only shade dependencies as a last resort | none |
| 19 | Place each package in only one module | none |
| 20 | Give each JAR file a module name | low |
| 21 | Upload artifacts to Maven Central |  |
| 22 | Declare all direct dependencies | low/medium |

**Comment by kenwenzel** [comment]
My take on how the Best Practices affect aas4j

| JLBP #  | Name | Effort | Notes |
| ------------- | ------------- | ------------- | ------------- |
| 9 | Support the minimum Java version of your consumers | low/none | Java 11 should stay the minimum nowadays
| 11 | Keep dependencies up to date | low/medium | use https://www.mojohaus.org/versions-maven-plugin/
| 20 | Give each JAR file a module name | low | maybe use bnd-maven-plugin to generate hybrid JPMS and OSGi modules
| 22 | Declare all direct dependencies | low/medium | use https://www.mojohaus.org/flatten-maven-plugin/

**Comment by mjacoby** [comment]
I'm not familiar with `versions-maven-plugin`. Is there any reason this is preferable to CI-based solutions like `dependabot`? I would prefer to be actively notified about new versions of a dependency which I guess is not possible with a maven plugin. 

And regarding **JLBP#22**, how does `flatten-maven-plugin` ensure the required properties are actually defined? Wouldn't be [maven-dependency-plugin](https://maven.apache.org/plugins/maven-dependency-plugin) the right tool?

**Comment by FrankSchnicke** [comment]
Here are my thoughts:
- **JLBP#2**: I think this an important point that we should consider before our first official release. Right now, every helper class (e.g., [JsonTreeProcessor](https://github.com/eclipse-aas4j/aas4j/blob/main/dataformat-json/src/main/java/io/adminshell/aas/v3/dataformat/json/modeltype/JsonTreeProcessor.java)) has to be considered part of our public API. Thus, any change to this class or its function signatures would be a breaking change.
- **JLBP#11**: I would also be in favor of dependabot due to its native integration in GitHub.

**Comment by in-fke** [comment]
I would also suggest to have self-contained artifactId that are "self-explaining"
e.g. instead of 
```
dataformat-aasx
dataformat-core
dataformat-json
dataformat-parent
dataformat-xml
model
```

have prefix "aas4j-" for every artifact.

```
aas4j-dataformat-aasx
aas4j-dataformat-core
aas4j-dataformat-json
aas4j-dataformat-parent
aas4j-dataformat-xml
aas4j-model
```

as ArtifactIds. Imagine the artifact "spring-core" would just be called "core" - no-one could relate that it's from Spring.

**Comment by FrankSchnicke** [comment]
Good idea! We discussed this yesterday in our committer meeting and agreed on this. Cf. https://github.com/eclipse-aas4j/aas4j/issues/195 for status and further details.
